src/datatype/svecf32.rs

```diff
@@ -0,0 +1,490 @@
+use crate::prelude::*;
+use pgrx::pg_sys::Datum;
+use pgrx::pg_sys::Oid;
+use pgrx::pgrx_sql_entity_graph::metadata::ArgumentError;
+use pgrx::pgrx_sql_entity_graph::metadata::Returns;
+use pgrx::pgrx_sql_entity_graph::metadata::ReturnsError;
+use pgrx::pgrx_sql_entity_graph::metadata::SqlMapping;
+use pgrx::pgrx_sql_entity_graph::metadata::SqlTranslatable;
+use pgrx::FromDatum;
+use pgrx::IntoDatum;
+use service::prelude::*;
+use std::alloc::Layout;
+use std::cmp::Ordering;
+use std::ffi::CStr;
+use std::ffi::CString;
+use std::ops::Deref;
+use std::ops::DerefMut;
+use std::ops::Index;
+use std::ops::IndexMut;
+use std::ptr::NonNull;
+
+#[repr(C, align(8))]
+pub struct SVecf32 {
+    varlena: u32,
+    len: u16,
+    kind: u8,
+    reserved: u8,
+    dims: u16,
+    padding: u16,
+    phantom: [SparseF32Element; 0],
+}
+
+impl SVecf32 {
+    fn varlena(size: usize) -> u32 {
+        (size << 2) as u32
+    }
+    fn layout(len: usize) -> Layout {
+        u16::try_from(len).expect("Vector is too large.");
+        let layout_alpha = Layout::new::<SVecf32>();
+        let layout_beta = Layout::array::<SparseF32Element>(len).unwrap();
+        let layout = layout_alpha.extend(layout_beta).unwrap().0;
+        layout.pad_to_align()
+    }
+    pub fn new_in_postgres(slice: &[SparseF32Element], dims: u16) -> SVecf32Output {
```

Comment:
We must **not** use `&[SparseF32Element]` everywhere. We must know `dims`. Use the type behind  `G::VectorRef<'a>`.

Reply:
completed in new commit